### PR TITLE
Basic port to 1.19.2

### PR DESCRIPTION
**Pull request by sisby-folk**
Hello again! Been a bit.

1.19.2 is mostly friendly, with a bit of syntax changes. 

Two major bumps along the way

 - StructureFeature is now StructureType and is slightly different, leading us to use `StructureType.JIGSAW` for the village marker, which is *okay* but bastions are gonna show up as funny little houses. Other option would be to flag each village type manually, but that doesn't work for modded villages. Maybe something APIs have solved.
 - Biome categories just.. aren't a thing anymore! You don't get biome information on init at all, which makes sense considering datapack biomes never had this (if i remember right). Anyway, this basically leads to us killing `guessFittingTextureSetFallback` (so it'll just use tags), and using tags instead where possible (like for priority for rivers, or checking whether something is nether). This means I've cached less, so if this is intensive that's worth a look.

Besides that, pretty smooth, had to do a few toolchain bumps but it worked out.

javaw_qX1BExzM4n
javaw_F3ESKwi0SS

### CODE REVIEW

**Comment by sisby-folk** [comment]
For your own interest, we also made an [itemless fork](https://github.com/sisby-folk/AntiqueAtlas/compare/1.19...sisby-folk:AntiqueAtlas:1.19_itemless) that we're planning to use for our no-modded-items modpack

**Comment by tyra314** [comment]
Thank you for the PR, I'll check it out.

> For your own interest, we also made an [itemless fork](https://github.com/sisby-folk/AntiqueAtlas/compare/1.19...sisby-folk:AntiqueAtlas:1.19_itemless) that we're planning to use for our no-modded-items modpack

No need to fork for that. Change `itemNeeded` config entry to false. If that doesn't work, patches are welcome.

**Comment by sisby-folk** [comment]
In this specific case (Tinkerer's Quilt) the entire modpack hinges on not adding items to the registry / world at all - it was fun to comb through all of that code though!

**Comment by sisby-folk** [comment]
Oop! extra thing - we've had this in our modpack for a hot minute and it
seems like tile choosing for nether isn't working, so we may have broken
that accidentally with a sensible-seeming replacement that didn't quite cut
it

**Comment by sisby-folk** [comment]
False alarm! I actually had `yttr` installed, which replaces nether biomes in some way - they show fine with vanilla!

**Comment by sisby-folk** [comment]
More notes as we've gone along. 

- Texture sets that need adding in 1.19 include:
   - mangrove_swamp
   - lush_caves
   - dripstone_caves
   - deep_dark
 - ~~We have no idea how to disable debug mode in builds (oops)~~ (oh, it's config)
 - We wrote a [fwaystones addon](https://github.com/sisby-folk/antique-fwaystones), which required manually mapping waystone hash strings to marker id ints (and saving the map persistently) - we might eventually ask about letting marker IDs be arbitrary or something.
 - While debugging, we've occasionally experienced S2C tile packets fail, causing filled atlases to get stuck blank. Seems timing and architectury related, but we can't reproduce it consistently. Yet to see it in a production environment like our pack.

```
java.lang.NullPointerException: Cannot invoke "net.minecraft.class_1657.method_5770()" because the return value of "dev.architectury.networking.NetworkManager$PacketContext.getPlayer()" is null 
   at hunternif.mc.impl.atlas.network.packet.s2c.play.MapDataS2CPacket.lambda$apply$0(MapDataS2CPacket.java:42) ~[transformed-mod-antiqueatlas.i0:0/:?]
```

**Comment by tyra314** [comment]
>     * While debugging, we've occasionally experienced S2C tile packets fail, causing filled atlases to get stuck blank. Seems timing and architectury related, but we can't reproduce it consistently. Yet to see it in a production environment like our pack.

Yeah, I've seen this shit as well. Really annoying. Didn´t notice though that it only occurred in dev environments. I could have sworn that I had a bug report for that.

File: common/src/main/java/hunternif/mc/impl/atlas/client/TileTextureMap.java [hunk 1]
```diff
@@ -97,6 +94,8 @@ static private Optional<Identifier> guessFittingTextureSet(RegistryKey<Biome> bi
     }
 
     static public Optional<Identifier> guessFittingTextureSetFallback(Biome biome) {
+        /*
```
Comment (tyra314) [inline common/src/main/java/hunternif/mc/impl/atlas/client/TileTextureMap.java]: If that doesn´t work anymore because Minecraft changed that, then there´s no reason to keep the code. In fact, the method should be removed instead of stitching useless stuff.

File: common/src/main/java/hunternif/mc/impl/atlas/client/gui/GuiAtlas.java [hunk 2]
```diff
@@ -1012,14 +1010,14 @@ public void render(MatrixStack matrices, int mouseX, int mouseY, float par3) {
...
                 String texture_set = TileTextureMap.instance().getTextureSet(tile).name.toString();
                 drawTooltip(Arrays.asList(
-                                new LiteralText(coords),
-                                new LiteralText(chunks),
-                                new LiteralText("Tile: " + tile),
-                                new LiteralText("TSet: " + texture_set)),
+                                Text.translatable(coords),
+                                Text.translatable(chunks),
+                                Text.translatable("Tile: " + tile),
+                                Text.translatable("TSet: " + texture_set)),
```
Comment (tyra314) [inline common/src/main/java/hunternif/mc/impl/atlas/client/gui/GuiAtlas.java]: LiteralText isn´t something that should be translated. Isn´t there a `Text.literal` in place for that?

File: common/src/main/java/hunternif/mc/impl/atlas/core/scanning/TileDetectorBase.java [hunk 3]
```diff
@@ -41,36 +47,54 @@ public class TileDetectorBase implements ITileDetector {
...
+    @ExpectPlatform
+    static private boolean hasSwampWater(RegistryEntry<Biome> biomeTag) {
+        throw new AssertionError("Not implemented");
     }
 
-    int priorityForBiome(Identifier biome) {
-        if (waterBiomes.contains(biome)) {
+    public static int priorityForBiome(World world, Biome biome) {
+        Registry<Biome> biomeRegistry = world.getRegistryManager().get(Registry.BIOME_KEY);
+        RegistryEntry<Biome> biomeTag = biomeRegistry.entryOf(biomeRegistry.getKey(biome).orElse(null));
```
Comment (tyra314) [inline common/src/main/java/hunternif/mc/impl/atlas/core/scanning/TileDetectorBase.java]: While I appreciate the new biome tags in general, I feel like getting to the `biomeTag` is such a pain in the ass. This makes me question whether this function get's the correct input.

File: common/src/main/java/hunternif/mc/impl/atlas/core/scanning/TileDetectorBase.java [hunk 4]
```diff
@@ -41,36 +47,54 @@ public class TileDetectorBase implements ITileDetector {
...
+//    }
+
+    @ExpectPlatform
+    static private boolean hasSwampWater(RegistryEntry<Biome> biomeTag) {
+        throw new AssertionError("Not implemented");
     }
 
-    int priorityForBiome(Identifier biome) {
-        if (waterBiomes.contains(biome)) {
+    public static int priorityForBiome(World world, Biome biome) {
```
Comment (tyra314) [inline common/src/main/java/hunternif/mc/impl/atlas/core/scanning/TileDetectorBase.java]: This can be private

File: common/src/main/java/hunternif/mc/impl/atlas/core/scanning/TileDetectorBase.java [hunk 5]
```diff
@@ -41,36 +47,54 @@ public class TileDetectorBase implements ITileDetector {
...
+//
+//    int priorityForBiome(Identifier biome) {
+//        if (waterBiomes.contains(biome)) {
+//            return 4;
+//        } else if (beachBiomes.contains(biome)) {
+//            return 3;
+//        } else {
+//            return 1;
+//        }
+//    }
```
Comment (tyra314) [inline common/src/main/java/hunternif/mc/impl/atlas/core/scanning/TileDetectorBase.java]: Don´t ever commit commented-out code.

File: common/src/main/java/hunternif/mc/impl/atlas/structure/Village.java [hunk 6]
```diff
@@ -1,13 +1,13 @@
...
-import net.minecraft.text.TranslatableText;
-import net.minecraft.world.gen.feature.StructureFeature;
+import net.minecraft.text.Text;
+import net.minecraft.world.gen.structure.StructureType;
 
 public class Village {
     public static void registerMarkers() {
         if (AntiqueAtlasMod.CONFIG.autoVillageMarkers) {
-            StructureHandler.registerMarker(StructureFeature.VILLAGE, AntiqueAtlasMod.id("village"), new TranslatableText("gui.antiqueatlas.marker.village"));
+            StructureHandler.registerMarker(StructureType.JIGSAW, AntiqueAtlasMod.id("village"), Text.translatable("gui.antiqueatlas.marker.village"));
```
Comment (tyra314) [inline common/src/main/java/hunternif/mc/impl/atlas/structure/Village.java]: This isn´t a good idea. Many mods will use the JIGSAW type; hence this may end up adding a gazillion village markers at places where some random mod generated stuff. I´d rather remove village markers entirely.

File: forge/src/main/java/hunternif/mc/impl/atlas/client/forge/TileTextureMapImpl.java [hunk 7]
```diff
@@ -23,7 +23,7 @@ static public Optional<Identifier> guessFittingTextureSet(RegistryKey<Biome> bio
 
         RegistryEntry<Biome> biomeTag = MinecraftClient.getInstance().world.getRegistryManager().get(Registry.BIOME_KEY).entryOf(biome);
 
-        if (biomeTag.isIn(Tags.Biomes.IS_END)) {
+        if (biomeTag.isIn(BiomeTags.IS_END)) {
```
Comment (tyra314) [inline forge/src/main/java/hunternif/mc/impl/atlas/client/forge/TileTextureMapImpl.java]: I'm not sure, if that's possible now, but does Fabric and Forge now have enough equivalent BiomeTags, so the biome-guessing can be harmonized between forge and fabric?

**Review by tyra314 — CHANGES_REQUESTED** [review]
Thanks for the work. There's quite a bit left, though.

Once it's ready, I'll create a new branch '1.19' as target.

File: common/src/main/java/hunternif/mc/impl/atlas/core/scanning/TileDetectorBase.java (hunk 3 above)
Comment (sleepingdragoninn) [reply common/src/main/java/hunternif/mc/impl/atlas/core/scanning/TileDetectorBase.java]: I've been playing around with the mod for a while and I'm definitely getting correct biomes on my overworld maps. So weird as it is, I think so? It's the same method used elsewhere in the code it just looks janky here.

File: forge/src/main/java/hunternif/mc/impl/atlas/client/forge/TileTextureMapImpl.java (hunk 7 above)
Comment (sleepingdragoninn) [reply forge/src/main/java/hunternif/mc/impl/atlas/client/forge/TileTextureMapImpl.java]: There's still an annoying spotty set. Things like swamps still don't have proper tags. What could be done is putting unified logic in one method, then having an impl-specific method that just returns an optional for biomes that don't have vanilla tags. LMK if you want us to give that a hand.

**Comment by sisby-folk** [comment]
@tyra314 on the "stuff failing to send to the client thing" - I think you might be right and it might be happening on prod too? I thought it was just a mod conflict, but I'm occasionally getting netty exceptions on both the client and server, and I've been experiencing a bug (which I thought was unrelated) where markers in the nether would get completely wiped. Not too sure there!

Here's the log bug that makes maps totally blank, though:

image

I think it happens consistently if you log out / save while holding an atlas in your hand. That would imply the game is checking for atlases before the player is actually initialized properly.

File: common/src/main/java/hunternif/mc/impl/atlas/client/TileTextureMap.java (hunk 1 above)
Comment (sleepingdragoninn) [reply common/src/main/java/hunternif/mc/impl/atlas/client/TileTextureMap.java]: Good point! replaced with an `Optional.empty()` expression instead in a61ff06

File: common/src/main/java/hunternif/mc/impl/atlas/client/gui/GuiAtlas.java (hunk 2 above)
Comment (sleepingdragoninn) [reply common/src/main/java/hunternif/mc/impl/atlas/client/gui/GuiAtlas.java]: Oops! must have been a find-replace issue. Went through the diff and fixed all of these in a61ff06

File: common/src/main/java/hunternif/mc/impl/atlas/core/scanning/TileDetectorBase.java (hunk 4 above)
Comment (sleepingdragoninn) [reply common/src/main/java/hunternif/mc/impl/atlas/core/scanning/TileDetectorBase.java]: used in TileDetectorNether! Updated to package-private in a61ff06

File: common/src/main/java/hunternif/mc/impl/atlas/core/scanning/TileDetectorBase.java (hunk 5 above)
Comment (sleepingdragoninn) [reply common/src/main/java/hunternif/mc/impl/atlas/core/scanning/TileDetectorBase.java]: Tell that to the rest of the commented code in the repo! (oops, assumed this was convention when porting by mistake - removed this in a61ff06)

File: common/src/main/java/hunternif/mc/impl/atlas/structure/Village.java (hunk 6 above)
Comment (sleepingdragoninn) [reply common/src/main/java/hunternif/mc/impl/atlas/structure/Village.java]: Great point! added a structure tag based map instead in a61ff06, those are lazy-loaded so it works a charm

image

File: common/src/main/java/hunternif/mc/impl/atlas/core/scanning/TileDetectorBase.java (hunk 3 above)
Comment (tyra314) [reply common/src/main/java/hunternif/mc/impl/atlas/core/scanning/TileDetectorBase.java]: You misunderstood me here. I was thinking about the method parameters. In particular, if `Biome` is the correct type, or if it should be a BiomeKey or something like that instead.

File: common/src/main/java/hunternif/mc/impl/atlas/core/scanning/TileDetectorBase.java (hunk 5 above)
Comment (tyra314) [reply common/src/main/java/hunternif/mc/impl/atlas/core/scanning/TileDetectorBase.java]: I try to get rid of that stuff, there's probably a lot lurking still.

**Comment by tyra314** [comment]
> I think it happens consistently if you log out / save while holding an atlas in your hand. That would imply the game is checking for atlases before the player is actually initialized properly.

Well, at least that is a good starting point to reproduce the issue.

File: forge/src/main/java/hunternif/mc/impl/atlas/client/forge/TileTextureMapImpl.java (hunk 7 above)
Comment (tyra314) [reply forge/src/main/java/hunternif/mc/impl/atlas/client/forge/TileTextureMapImpl.java]: I would love to see the general algorithm shared between forge and fabric and only the different tags impl-specific, but I'm unsure if that is feasible.

File: common/src/main/java/hunternif/mc/impl/atlas/core/scanning/TileDetectorBase.java (hunk 3 above)
Comment (sleepingdragoninn) [reply common/src/main/java/hunternif/mc/impl/atlas/core/scanning/TileDetectorBase.java]: You might be getting mappings mixed up - in yarn it's `Biome` - the closest thing to `BiomeKey` is `BiomeKeys`, which is used to register and define all the Vanilla biomes to the BIOME_KEY registry (similar to BiomeTags or Items)

File: forge/src/main/java/hunternif/mc/impl/atlas/client/forge/TileTextureMapImpl.java (hunk 7 above)
Comment (sleepingdragoninn) [reply forge/src/main/java/hunternif/mc/impl/atlas/client/forge/TileTextureMapImpl.java]: Smashed up and done in 8b7f924 - the set priority is now the same between platforms, and it should be easier to see how redundant or lacking tag coverage is.

File: common/src/main/java/hunternif/mc/impl/atlas/core/scanning/TileDetectorBase.java (hunk 3 above)
Comment (tyra314) [reply common/src/main/java/hunternif/mc/impl/atlas/core/scanning/TileDetectorBase.java]: Let me rephrase.  What hits me is the over-and-over repeated code to get a biomeTag from a biome.
In particular, this method gets called from `getBiomeID`. At call time, there already is the biomeTag available, but the biome is passed instead. Leading to this method having a duplication of code and being a bit less efficient because it has to do the lookup again. Having said that, from what the method name suggests, one would not expect to pass in a biomeTag instance, but a little rename would fix that.

Comment (sleepingdragoninn) [reply common/src/main/java/hunternif/mc/impl/atlas/core/scanning/TileDetectorBase.java]: Yep, got it, that's on me - the tiny diff github was giving me here was leading me astray, swapped to just `RegistryEntry<Biome>` instead of passing in both world and biome in 733485f

**Comment by Daenyth** [comment]
Is there any way I can help with this?

**Comment by sisby-folk** [comment]
@tyra314 hi hello! I'm at the point where I'm thinking about poking 1.20.1 with a stick, any further thoughts on this one? (besides not being able to display underground biomes, naturally)

**Comment by Daniilnm** [comment]
hello, I want to ask how the work is progressing, there have been no changes in this branch for a long time

**Comment by Daniilnm** [comment]
> @tyra314 hi hello! I'm at the point where I'm thinking about poking 1.20.1 with a stick, any further thoughts on this one? (besides not being able to display underground biomes, naturally)

.

**Comment by tadhgmister** [comment]
@sisby-folk, I tried using the artifact from this pull request but ran into an error:
```
java.lang.IllegalAccessError: 
class hunternif.mc.impl.atlas.client.TileTextureMap tried to access private method
'boolean hunternif.mc.impl.atlas.client.forge.TileTextureMapImpl.biomeIsJungle(net.minecraft.core.Holder)'
```
And looking at the [file, `biomeIsJungle` is the only private method of the lot,](https://github.com/sisby-folk/AntiqueAtlas/blob/1.19/forge/src/main/java/hunternif/mc/impl/atlas/client/forge/TileTextureMapImpl.java#L44) every other one is public and there was [a commit 3 months ago where you fixed visibility for several other methods](https://github.com/sisby-folk/AntiqueAtlas/commit/d8b8ffc59d496114bcb04484567589e1989b2689) but jungle was left out.

It crashes as soon as I try to load a world using forge, I'm not sure how to suggest an edit, do you think you could fix that in the correct place to change it?
Thanks :)

File: forge/src/main/java/hunternif/mc/impl/atlas/client/forge/TileTextureMapImpl.java [hunk 8]
```diff
@@ -1,167 +1,99 @@
...
+
+    public static boolean biomeIsIcy(RegistryEntry<Biome> biomeTag) {
+        return biomeTag.isIn(Tags.Biomes.IS_COLD) || biomeTag.isIn(Tags.Biomes.IS_SNOWY);
+    }
+
+    public static boolean biomeIsShore(RegistryEntry<Biome> biomeTag) {
+        return biomeTag.isIn(BiomeTags.IS_BEACH);
+    }
+
+    private static boolean biomeIsJungle(RegistryEntry<Biome> biomeTag) {
```
Comment (tadhgmister) [inline forge/src/main/java/hunternif/mc/impl/atlas/client/forge/TileTextureMapImpl.java]:
```suggestion
    public static boolean biomeIsJungle(RegistryEntry<Biome> biomeTag) {
```

**Review by tadhgmister — CHANGES_REQUESTED** [review]
crashes for me immidiately upon joining a world in forge because it needs this to be public, I assume this was an oversight.

File: forge/src/main/java/hunternif/mc/impl/atlas/client/forge/TileTextureMapImpl.java (hunk 8 above)
Comment (sleepingdragoninn) [reply forge/src/main/java/hunternif/mc/impl/atlas/client/forge/TileTextureMapImpl.java]: sorted in ec01126

**Comment by TheOverpassArsonist** [comment]
What's the status of this? As far as I can tell issue that's been mentioned has already had commits addressing it/them, and every item in the requested changes by tyra show as resolved. Is this done and it's just a matter of publishing it, or are there remaining issues? (for reference, not a mod developer myself, but I checked in to see what the progress was at and I noticed that it seems like everything that's been mentioned has been addressed)

The one thing that github is citing as an issue is the requested changes, but as far as I can tell all of the requested changes have been made. On sept 11th [this](https://github.com/AntiqueAtlasTeam/AntiqueAtlas/pull/467#issuecomment-1713159362) comment was made implying it was solid about a month and a half ago and (aside from one private/public mixup) it doesn't look like any other issues have cropped up.

Not trying to poke or prod for some ETA/date just curious as to the general state this PR is in since it appears as though every fix that was requested/needed has been made already.

**Comment by sisby-folk** [comment]
Hey hello! @tyra314 just a heads up that we're planning to put our itemless forks of this PR (and the 1.20.1 one) on modrinth within a week or so. We weren't originally planning on it, but someone else went and did it without asking or notifying us, and we'd like to avoid that happening a second time. 

We'll be forgoing any monetization on modrinth and crediting clearly to suit the cc-nc-by-sa textures.

Naturally, let us know if there are any issues, and as always you're still perfectly welcome to adapt any of our changes in these PRs into the main mod and MR project. Cheers!
